### PR TITLE
internal/dag: Implement GatewayAllow for TLSRoutes

### LINKED ISSUE #3695: Implement GatewayAllow in TLSRoute
Is [this block of logic](https://github.com/projectcontour/contour/blob/e3cf805af778fe42ec39681dbf71b783b091007e/internal/dag/gatewayapi_processor.go#L138-L157) that we have for `HTTPRoutes` not needed for `TLSRoutes`?

_Originally posted by @skriss in https://github.com/projectcontour/contour/pull/3627#discussion_r634681222_

### PR DESCRIPTION
**Pull request by stevesloka**
Fixes #3695

Signed-off-by: Steve Sloka <slokas@vmware.com>

### CODE REVIEW

**Comment by stevesloka** [comment]
I'm going to follow up with some integration tests which would be nice to route through the API server properly.

File: internal/dag/builder_test.go [hunk 1]
```diff
@@ -707,7 +707,7 @@ func TestDAGInsertGatewayAPI(t *testing.T) {
 					},
 					Spec: gatewayapi_v1alpha1.HTTPRouteSpec{
 						Gateways: &gatewayapi_v1alpha1.RouteGateways{
-							Allow: gatewayAllowTypePtr(gatewayapi_v1alpha1.GatewayAllowAll),
+							Allow: gatewayAllowTypePtr(gatewayapi_v1alpha1.GatewayAllowSameNamespace),
```
Comment (stevesloka) [inline internal/dag/builder_test.go:710]: This was wrong previously and should have been "SameNamespace". The result it still correct, but just fixing up in this PR.

Comment (stevesloka) [reply internal/dag/builder_test.go:710]: Also I think we could DRY this tests up, there's a lot of duplication. 😢

**Comment by stevesloka** [comment]
My logic is broken somewhere causing the TLSRoute to be kicked out. Digging into that now.

**Review by skriss — APPROVED** [review]
LGTM

**Comment by stevesloka** [comment]
I didn't update the _integration tests, probably need fixes. Do we care at this point since we're trashing those?

**Comment by sunjayBhatia** [comment]
> I didn't update the _integration tests, probably need fixes. Do we care at this point since we're trashing those?

idk if the _integration removal will make it in before 1.16 so just to keep the pipeline green, maybe fix or remove the test assuming it is mirrored into e2e

**Review by sunjayBhatia — COMMENTED** [review]
LGTM pending CI (removal of _integration test or adding missing allow all selector)

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**